### PR TITLE
Reintroduce a fix for strongbox decryption data corruption (#383)

### PR DESCRIPTION
**Pull request by cshfang**
A PR (https://github.com/oblador/react-native-keychain/pull/288) a while back was merged to fix an issue where decrypting the keychain using a `CipherInputStream` causes certain Google Pixel devices to throw an exception and hang.

The change seems to have been inadvertently reverted in the large refactoring that took place in https://github.com/oblador/react-native-keychain/pull/260.

This PR reintroduces that fix.

### CODE REVIEW

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java
Comment (aeirola) [inline android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java:359]:
```suggestion
        handler.initialize(cipher, key, bytes);
```

Now that the whole decrypt process works with byte arrays, it might be better to change the `DecryptBytesHandler` to take the input as a byte array, instead of an input stream.

This change also requires the `DecryptBytesHandler` interface input parameter to be changed to `byte[]`, but shouldn't require other changes as the parameter is only passed to overloaded methods which already handle byte arrays as well as InputStreams.

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java
Comment (aeirola) [inline android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java:353]: Might be good to add a comment in the code with a link to the issue, or this PR, describing why we don't use input streams. Otherwise someone might be eager to refactor this again, without testing on a pixel device first :)

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java
Comment (aeirola) [inline android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java:355]:
```suggestion
    try {
```

I don't think the input stream and output stream are really used anymore, since the decrypt operation now works with byte arrays. These should probably be removed.

**Review by aeirola — CHANGES_REQUESTED** [review]
In general, the code looks good! Would still like to clean it up a bit by removing the mostly unused byte array streams. Otherwise I'm worried they might cause some confusion in the future.

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java
Comment (cshfang) [reply android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java:353]: No problem at, makes complete sense to me.

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java
Comment (cshfang) [reply android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java:359]: 👍 This was a miss. I should've cleaned this up.

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java
Comment (cshfang) [reply android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java:355]: 👍 This was a miss. I should've cleaned this up.

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java
Comment (cshfang) [inline android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java:529]: When switching over to this overloaded method handling `byte[]`, this condition was triggered. Reading through this method, I believe the intent may have been to validate that the input bytes array is actually *greater* than `IV_LENGTH`.

Comment (aeirola) [reply android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java:529]: Whoa, strange bug. Wonder if this code was ever used before.

**Comment by aeirola** [comment]
The code looks good, and normal encoding and decoding seems to work OK on Android. But when trying to get stored values which are protected with fingerprint authentication, I get the a `javax.crypto.IllegalBlockSizeException` error caused by `android.security.KeyStoreException: Invalid input length`. Apparently the AES and RSA cipher storages use the same `decryptBytes` method, and the RSA cipher used for fingerprints doesn't allow decyrpting data using `doFinal` with arbitrary length input.

Might be that we need to move the modified `decryptBytes` method from `CipherStorageBase` to the `CipherStorageKeystoreAesCbc` class, so that we don't interfere with the workings of the other cipher storage implementations.

This issue wasn't present in the previous fix for the same issue, as the Android fingerprints weren't supported yet at the time.

**Comment by cshfang** [comment]
> The code looks good, and normal encoding and decoding seems to work OK on Android. But when trying to get stored values which are protected with fingerprint authentication, I get the a `javax.crypto.IllegalBlockSizeException` error caused by `android.security.KeyStoreException: Invalid input length`. Apparently the AES and RSA cipher storages use the same `decryptBytes` method, and the RSA cipher used for fingerprints doesn't allow decyrpting data using `doFinal` with arbitrary length input.
> 
> Might be that we need to move the modified `decryptBytes` method from `CipherStorageBase` to the `CipherStorageKeystoreAesCbc` class, so that we don't interfere with the workings of the other cipher storage implementations.
> 
> This issue wasn't present in the previous fix for the same issue, as the Android fingerprints weren't supported yet at the time.

Ah thanks for the good catch. Based on what you're saying, it makes sense to me to try to make the modified `decryptBytes` method the exception for AES rather than the norm. I'll work on this today.

**Comment by cshfang** [comment]
@aeirola I pushed a refactor to override `decryptBytes` within the `CipherStorageKeystoreAesCbc` class. It seems like the changes are still holding up in my testing but I only tested with a fairly contrived RSA set/get. I'm not sure how to replicate the `android.security.KeyStoreException: Invalid input length` you found above so may have to lean on you for further testing.

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java
Comment (aeirola) [inline android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java:557]: I'm wondering if making this interface generic causes some confusion, especially when the different `CipherStorage` implementations don't explicitly state which type of `DecryptBytesHandler` they expect. Maybe would be better to not touch these common files too much, and just omit using the `handler` in the overridden `CipherStorageKeystoreAesCbc.decryptBytes`.

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreAesCbc.java
Comment (aeirola) [inline android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreAesCbc.java:237]:
```suggestion
      final IvParameterSpec iv = IV.readIv(bytes);
      cipher.init(Cipher.DECRYPT_MODE, key, iv);
```

In order to not have to make the handler interface generic.

I realise that this will make the implementation deviate a bit from the other `CipherStorage` classes, but probably best to keep this workaround as contained as possible.

**Review by aeirola — APPROVED** [review]
Code looks good, and works well in my testing. Might personally prefer to keep the changes even more local to the `CipherStorageKeystoreAesCbc` class by not using generics for the `DecryptBytesHandler` interface, but this is more of a matter of taste. Feel free to change it or ignore my comments :)

Before merging I'll try to contact people who've faced the issue and see wether they would be willing to help out testing this PR with actual Titan M devices.

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java
Comment (cshfang) [reply android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java:557]: Sure, I understand the concern. I considered just adding comments to specify the generics handled but didn't want to pigeonhole the interface too much. Also opted to not stray from the original implementation too much and therefore kept the `handler` but your suggestion makes sense also and I'm happy to make that change.

**Comment by cshfang** [comment]
> Before merging I'll try to contact people who've faced the issue and see wether they would be willing to help out testing this PR with actual Titan M devices.

FWIW, I have been testing this against my physical Pixel 4 device as well, which I believe utilizes the Titan M.

**Comment by aeirola** [comment]
@Metroidaron @nicolashemonic Hi, I remember that you've previously helped this project by testing issues on Pixel devices in #208 . Now we're in a situation where we have faced a regression in the support for Pixel devices with Titan M security chips. In order to verify that our fix is valid, I'd like to ask wether you would be able to help out once again by checking if the latest code in this PR still works on your devices and use cases. The code from this PR can be installed with `npm install github:cshfang/react-native-keychain#fix-strongbox-decryption`. Thanks again!

**Comment by aeirola** [comment]
Unfortunately there wasn't much interest in testing the fix on additional devices. Would have been nice with some extra validations with different devices, but won't let that block the deployment of the fix as the code looks good and you already tested on one Pixel 4 device. I'll merge this and see if we can get it released soon.

**Comment by aeirola** [comment]
@oblador Do we have a schedule for the next release? Releaseing this fix would be useful for a certain group of Android users.

**Comment by oblador** [comment]
@aeirola It's released in 6.2.0!

**Comment by nicolashemonic** [comment]
Hi @aeirola @oblador 

Sorry for the silence because I was in long vacation 🏖 and busy when back to the office.

I upgraded my app to `react-native-keychain@6.2.0` and I confirm that the fix works well on Pixel 3a Android 11. The authentication (very long token stored in keychain) is not lost when the app is closed and opened.

Thanks !

**Pull request merged**
